### PR TITLE
Feature http_tcp_listeners - Add redirect default action type

### PR DESCRIPTION
**Pull request by gaelreyrol**
Hi,

With the actual version the module only allow forward listeners type. Recently AWS added redirect listeners type. For example we can now redirect HTTP to HTTPS directly from lb resources.

I seperated `aws_lb_listeners` by default action type, one for forward and one for redirect but only on http_tcp listeners.

- Before `frontend_http_tcp`
- After `frontend_http_tcp_forward` & `frontend_http_tcp_redirect`

It breaks actual use of the module since the resource `aws_lb_listener.frontend_http_tcp` doesn't exist anymore.

Documentation have been updated as well as tests.
Terraform AWS provider `1.33.0` is required to use this feature.

### CODE REVIEW

**Comment by neoshadybeat** [comment]
I hope this PR will be merged soon

**Comment by colindean** [comment]
I would very much like this to be merged!

File: alb_no_logs.tf [hunk 1]
```diff
@@ -54,18 +54,51 @@ resource "aws_lb_target_group" "main_no_logs" {
...
+  port              = "${lookup(var.http_tcp_listeners_redirect[count.index], "port")}"
+  protocol          = "${lookup(var.http_tcp_listeners_redirect[count.index], "protocol")}"
+  count             = "${var.logging_enabled ? 0 : var.http_tcp_listeners_redirect_count}"
+
+  default_action {
+    target_group_arn = "${aws_lb_target_group.main_no_logs.*.id[lookup(var.http_tcp_listeners_redirect[count.index], "target_group_index", 0)]}"
+    type             = "redirect"
+
+    redirect {
+      host        = "${lookup(var.http_tcp_listeners_redirect[count.index], "redirect_host", lookup(var.default_action_redirect_defaults, "host"))})}"
```
Comment (smelchior) [inline alb_no_logs.tf]: There is a duplicate (} in here, so terraform will try to create with these properties:
`default_action.0.redirect.0.host:        "" => "#{host})}"`

Comment (colindean) [inline alb_no_logs.tf]:
```suggestion
      host        = "${lookup(var.http_tcp_listeners_redirect[count.index], "redirect_host", lookup(var.default_action_redirect_defaults, "host"))}""
```

Here's a suggestion to fix it based https://github.com/terraform-aws-modules/terraform-aws-alb/commit/de6fe240dead819a1a9ce9d4fe9fdc7223ebe5f5 that's included in [my fork](https://github.com/arcadia/terraform-aws-alb).

**Comment by Zevran** [comment]
Any news ?

**Comment by okonon** [comment]
@colindean has this been considered again ?

**Comment by colindean** [comment]
@okonon: It's in the fork linked above but I'm no longer associated with that organization and am thus no longer responsible for that fork.

**Comment by okonon** [comment]
@alkalinecoffee maybe you can help?

**Comment by Zevran** [comment]
It should be ok now @brandoconnor !

**Comment by okonon** [comment]
@brandoconnor @Zevran seems to be a problem with travis build

**Comment by okonon** [comment]
Anyone can fix and merge?

**Comment by RiceBowlJr** [comment]
Any updates on the build fail? I'd love use this feature :)

**Comment by danvaida** [comment]
While I think this will be useful for many people (if/when it'll be merged), I wanted to take it a step further and add support for multiple types of rules in combination with different default actions.
For example, having an HTTPS listener with 5 redirect rules and a default action of fixed-response. Another variation would be to have some redirect rules and forwarding as the default action. The list of potential combination goes on.
That kind of functionality requires the usage of the `aws_lb_listener_rule` and in combination with the severe limitations of TF `0.11`, it's simply more straightforward to go without a module or build one with a very specific purpose.

**Comment by fmunteanu** [comment]
@brandoconnor do you think is possible to review and merge the changes?

**Comment by aidan-melen** [comment]
This PR would be awesome. Here is a simple work around until this is merged:

```
module "alb" {
  source  = "terraform-aws-modules/alb/aws"
  version = "~> 4.0"

  load_balancer_name            = "my-alb"
  security_groups               = ["sg-edcd9784", "sg-edcd9785"]
  log_bucket_name               = "logs-us-east-2-123456789012"
  log_location_prefix           = "my-alb-logs"
  subnets                       = ["subnet-abcde012", "subnet-bcde012a"]
  tags                          = "${map("Environment", "test")}"
  vpc_id                        = "vpc-abcde012"
  https_listeners               = "${list(map("certificate_arn", "arn:aws:iam::123456789012:server-certificate/test_cert-123456789012", "port", 443))}"
  https_listeners_count         = "1"
  http_tcp_listeners            = "${list(map("port", "80", "protocol", "HTTP"))}"
  http_tcp_listeners_count      = "1"
  target_groups                 = "${list(map("name", "foo", "backend_protocol", "HTTP", "backend_port", "80"))}"
  target_groups_count           = "1"
}

resource "aws_lb_listener" "front_end_http_redirect" {
  load_balancer_arn = "${module.alb.load_balancer_id}"
  port              = "80"
  protocol          = "HTTP"

  default_action {
    type = "redirect"

    redirect {
      port        = "443"
      protocol    = "HTTPS"
      status_code = "HTTP_301"
    }
  }
}
```

**Comment by detkin** [comment]
@aidan-melen thanks for the suggestion. When I create an aws_lb_listener terraform creates the resource but it looks like the default created by the module, the forward rule, is the one that actually gets set on the ALB.

Am I missing something? Is there a way to get the redirect listener to be the default?

**Comment by aidan-melen** [comment]
> @aidan-melen thanks for the suggestion. When I create an aws_lb_listener terraform creates the resource but it looks like the default created by the module, the forward rule, is the one that actually gets set on the ALB.
> 
> Am I missing something? Is there a way to get the redirect listener to be the default?

well we typically want out alb to have two listeners: https forward to targets and http redriect to https. Currently, This alb module allows us to configure multiple https forward listener, but it does not allow for http redirect to https listeners. Thus the work around is the create a separate resource listener.

**Comment by detkin** [comment]
> > @aidan-melen thanks for the suggestion. When I create an aws_lb_listener terraform creates the resource but it looks like the default created by the module, the forward rule, is the one that actually gets set on the ALB.
> > Am I missing something? Is there a way to get the redirect listener to be the default?
> 
> well we typically want our alb to have two listeners: https forward to targets and http redriect to https listener. Currently, This alb module allows us to configure multiple https forward listener, but it does not allow for http redirect to https listeners. Thus the work around is to create a separate resource listener.

Thanks for the quick response. I understand that we're creating a separate resource listener and that we're setting it as the default action. It seems like a great solution to work around the limitation in the module. However, when I include a similar definition, it isn't attached the ALB created by the module. I only see the default forward rule created by the module for HTTP, not the redirect rule as we'd like to see.

Is there something missing from your example that I'd need to include to include the separate resource listener as an HTTP listener on the ALB? Does `http_tcp_listeners_count` need to be 0 and not defined in the module?

**Comment by aidan-melen** [comment]
> > > @aidan-melen thanks for the suggestion. When I create an aws_lb_listener terraform creates the resource but it looks like the default created by the module, the forward rule, is the one that actually gets set on the ALB.
> > > Am I missing something? Is there a way to get the redirect listener to be the default?
> > 
> > 
> > well we typically want our alb to have two listeners: https forward to targets and http redriect to https listener. Currently, This alb module allows us to configure multiple https forward listener, but it does not allow for http redirect to https listeners. Thus the work around is to create a separate resource listener.
> 
> Thanks for the quick response. I understand that we're creating a separate resource listener and that we're setting it as the default action. It seems like a great solution to work around the limitation in the module. However, when I include a similar definition, it isn't attached the ALB created by the module. I only see the default forward rule created by the module for HTTP, not the redirect rule as we'd like to see.
> 
> Is there something missing from your example that I'd need to include to include the separate resource listener as an HTTP listener on the ALB? Does `http_tcp_listeners_count` need to be 0 and not defined in the module?

as long as the alb is create with HTTPS forward listeners, you can get the id of the ALB from the module and pass it to the listener resource configured for HTTP to HTTPS redirect or whatever your want there.

http_tcp_listeners_count listener count is used to specify the number of HTTP forward listeners that the alb module will manage and NOT HTTP redirect listeners. if you are doing a HTTP to HTTPS redirect, then you will not want a HTTP forward listener. Hope that helps.

**Comment by aidan-melen** [comment]
> http_tcp_listeners_count

awe i see your confusion. I removed the `http_tcp_listeners_count` from the example above. hope that helps

**Comment by detkin** [comment]
> > http_tcp_listeners_count
> 
> awe i see your confusion. I removed the `http_tcp_listeners_count` from the example above. hope that helps

That worked perfectly, thanks for the help.

**Comment by Xat59** [comment]
@colindean , @smelchior , @brandonjbjelland 

Any news on it ? The requested feature would be really useful !

It seems the failed travis was caused because of the non definition of the AWS_REGION :
`export AWS_REGION=$(docker run --env AWS_DEFAULT_REGION=us-east-2 --env AWS_SECRET_ACCESS_KEY=${AWS_SECRET_ACCESS_KEY} --env AWS_ACCESS_KEY_ID=${AWS_ACCESS_KEY_ID} garland/aws-cli-docker aws ec2 describe-regions --query 'Regions[].{Name:RegionName}' --output text | shuf | head -n1)
` 

So AWS_REGION environment variable is empty and causes the fail.

**Comment by luckytaxi** [comment]
Any updates to this?

**Comment by Dmitry1987** [comment]
does the module support HTTP->HTTPS redirect right now? The example documentation says it is, but it throws an error. I guess I'll open a new issue for it.

**Comment by antonbabenko** [comment]
Hi all,

Forgot to mention that all types of ALB listener actions (redirect, fixed-response, authenticate-cognito, authenticate-oidc) are supported since [v5.3.0](https://github.com/terraform-aws-modules/terraform-aws-alb/releases/v5.3.0) of this module.

Check out code in [examples/complete-alb](https://github.com/terraform-aws-modules/terraform-aws-alb/tree/master/examples/complete-alb) for more details.

Closing this issue.

**Pull request closed without merging**